File: ros2-for-beginners/ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from functools import partial

from example_interfaces.srv import AddTwoInts


class AddTwoIntClient(Node):
    def __init__(self) -> None:
        super().__init__('add_two_ints_client')
        self.get_logger().info("python add two int client started")

        # call the function to invoke server functionalities
        self.call_add_two_ints_server()

    # function to call the server
    def call_add_two_ints_server(self):
        # arg: service_type,service_name
        client = self.create_client(AddTwoInts, "add_two_ints")

        # wait for server
        while not client.wait_for_service(1.0):
            self.get_logger().warn("waiting for add two ints server . . .")

        self.get_logger().info("<-- server available -->")

        # create a request
        a = int(input('enter number 1 : '))
        b = int(input('enter number 2 : '))
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        # call async and get it to future
        future = client.call_async(request)

        # add callback
        future.add_done_callback(partial(self.callback_add_two_ints, a=a, b=b))

    def callback_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f"{a} + {b} = {response.sum}")
        except Exception as e:
            self.get_logger().info(f"service call failer with exception : {e}")


def main(args=None):
    rclpy.init()
    node = AddTwoIntClient()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2-for-beginners/ros2_ws/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
#include <cstdlib>

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode(int a, int b, int argc) : Node("add_two_ints_client")
    {
        RCLCPP_INFO(this->get_logger(), "cpp add two ints client started");

        // we need to call the callback from different frame / thread inorder for function to not block the class constructor
        thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsServie, this, a, b));
    }

    void callAddTwoIntsServie(int a, int b)
    {
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_INFO(this->get_logger(), "waiting for server to be up ...");
        }

        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        auto future = client->async_send_request(request);
        /*wait until the future has a response values and can handle exception*/

        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %ld ", a, b, response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "service call failed");
        }
    }

private:
    //  create a new thread
    std::thread thread1_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    if (argc != 3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "usage : add_two_ints_client A B");
    }
    auto node = std::make_shared<AddTwoIntsClientNode>(atoll(argv[1]), atoll(argv[2]), argc);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}